packages/@uppy/core/src/Uppy.js

```diff
@@ -417,8 +451,11 @@ class Uppy {
     const { allowNewUpload } = this.getState()
 
     if (allowNewUpload === false) {
-      const error = new RestrictionError(this.i18n('noMoreFilesAllowed'))
-      this.#informAndEmit(error, file)
+      const error = file != null
+        ? new FileRestrictionError(this.i18n('noMoreFilesAllowed'), { file })
+        : new RestrictionError(this.i18n('noMoreFilesAllowed'))
```

Comment:
Why do we need a new error class for this?

Reply:
maybe we don't need it, but i thought it was more intention revealing. i guess we could also just have an optional `file` option for RestrictionError